File: ios/VibeArcade/Core/Networking/WebSocketService.swift
import Foundation
import Combine

class WebSocketService: NSObject, ObservableObject {
    static let shared = WebSocketService()

    private var webSocketTask: URLSessionWebSocketTask?
    private var urlSession: URLSession?

    @Published var connectionState: WebSocketConnectionState = .disconnected
    @Published var lastError: Error?

    private let messageSubject = PassthroughSubject<WebSocketMessage, Never>()
    var messagePublisher: AnyPublisher<WebSocketMessage, Never> {
        messageSubject.eraseToAnyPublisher()
    }

    private override init() {
        super.init()
        setupURLSession()
    }

    private func setupURLSession() {
        let config = URLSessionConfiguration.default
        urlSession = URLSession(configuration: config, delegate: self, delegateQueue: nil)
    }

    // MARK: - Connection Management
    func connect() {
        guard connectionState != .connected && connectionState != .connecting else {
            return
        }

        guard let token = AuthManager.shared.accessToken else {
            lastError = WebSocketError.noAuthToken
            return
        }

        connectionState = .connecting

        var request = URLRequest(url: URL(string: APIConfiguration.shared.websocketURL)!)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        webSocketTask = urlSession?.webSocketTask(with: request)
        webSocketTask?.resume()

        startListening()
    }

    func disconnect() {
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        connectionState = .disconnected
    }

    // MARK: - Message Handling
    private func startListening() {
        webSocketTask?.receive { [weak self] result in
            switch result {
            case .success(let message):
                self?.handleReceivedMessage(message)
                self?.startListening() // Continue listening
            case .failure(let error):
                self?.handleError(error)
            }
        }
    }

    private func handleReceivedMessage(_ message: URLSessionWebSocketTask.Message) {
        switch message {
        case .string(let text):
            if let data = text.data(using: .utf8),
               let webSocketMessage = try? JSONDecoder().decode(WebSocketMessage.self, from: data) {
                DispatchQueue.main.async {
                    self.messageSubject.send(webSocketMessage)
                }
            }
        case .data(let data):
            if let webSocketMessage = try? JSONDecoder().decode(WebSocketMessage.self, from: data) {
                DispatchQueue.main.async {
                    self.messageSubject.send(webSocketMessage)
                }
            }
        @unknown default:
            break
        }
    }

    private func handleError(_ error: Error) {
        DispatchQueue.main.async {
            self.lastError = error
            self.connectionState = .disconnected
        }
    }

    // MARK: - Send Messages
    func send(message: WebSocketMessage) {
        guard connectionState == .connected else {
            lastError = WebSocketError.notConnected
            return
        }

        do {
            let data = try JSONEncoder().encode(message)
            let string = String(data: data, encoding: .utf8)!
            let message = URLSessionWebSocketTask.Message.string(string)

            webSocketTask?.send(message) { [weak self] error in
                if let error = error {
                    DispatchQueue.main.async {
                        self?.lastError = error
                    }
                }
            }
        } catch {
            lastError = error
        }
    }

    // MARK: - Convenience Methods
    func joinRoom(_ roomId: String) {
        let message = WebSocketMessage(
            type: .joinRoom,
            roomId: roomId,
            playerId: AuthManager.shared.currentUser?.id ?? UUID(),
            data: nil,
            timestamp: Date()
        )
        send(message: message)
    }

    func leaveRoom(_ roomId: String) {
        let message = WebSocketMessage(
            type: .leaveRoom,
            roomId: roomId,
            playerId: AuthManager.shared.currentUser?.id ?? UUID(),
            data: nil,
            timestamp: Date()
        )
        send(message: message)
    }

    func sendGameMove(roomId: String, moveData: Data) {
        let message = WebSocketMessage(
            type: .gameMove,
            roomId: roomId,
            playerId: AuthManager.shared.currentUser?.id ?? UUID(),
            data: moveData,
            timestamp: Date()
        )
        send(message: message)
    }

    func sendChatMessage(roomId: String, text: String) {
        let chatData = ChatData(message: text)
        if let data = try? JSONEncoder().encode(chatData) {
            let message = WebSocketMessage(
                type: .chatMessage,
                roomId: roomId,
                playerId: AuthManager.shared.currentUser?.id ?? UUID(),
                data: data,
                timestamp: Date()
            )
            send(message: message)
        }
    }
}

// MARK: - URLSessionWebSocketDelegate
extension WebSocketService: URLSessionWebSocketDelegate {
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
        DispatchQueue.main.async {
            self.connectionState = .connected
        }
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        DispatchQueue.main.async {
            self.connectionState = .disconnected
        }
    }
}

// MARK: - Supporting Types
enum WebSocketConnectionState {
    case disconnected
    case connecting
    case connected
}

enum WebSocketError: Error {
    case noAuthToken
    case notConnected
    case encodingError
    case decodingError

    var localizedDescription: String {
        switch self {
        case .noAuthToken:
            return "No authentication token available"
        case .notConnected:
            return "WebSocket not connected"
        case .encodingError:
            return "Failed to encode message"
        case .decodingError:
            return "Failed to decode message"
        }
    }
}

struct ChatData: Codable {
    let message: String
}

File: android/app/src/main/java/com/acmegames/vibearcade/core/networking/WebSocketService.kt
package com.acmegames.vibearcade.core.networking

import com.acmegames.vibearcade.core.models.WebSocketMessage
import com.acmegames.vibearcade.core.storage.AuthManager
import com.google.gson.Gson
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WebSocketService @Inject constructor(
    private val authManager: AuthManager
) {
    private val gson = Gson()
    private var webSocket: WebSocket? = null
    private val messagesChannel = Channel<WebSocketMessage>(Channel.UNLIMITED)

    private val _connectionState = MutableStateFlow(WebSocketConnectionState.DISCONNECTED)
    val connectionState: StateFlow<WebSocketConnectionState> = _connectionState.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()

    val messages: Flow<WebSocketMessage> = messagesChannel.receiveAsFlow()

    private val client = OkHttpClient.Builder().build()

    suspend fun connect() {
        if (_connectionState.value == WebSocketConnectionState.CONNECTED ||
            _connectionState.value == WebSocketConnectionState.CONNECTING) {
            return
        }

        val token = authManager.getAccessToken()
        if (token.isNullOrEmpty()) {
            _lastError.value = "No authentication token available"
            return
        }

        _connectionState.value = WebSocketConnectionState.CONNECTING

        val request = Request.Builder()
            .url(ApiConfiguration.WEBSOCKET_URL)
            .header("Authorization", "Bearer $token")
            .build()

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                _connectionState.value = WebSocketConnectionState.CONNECTED
                _lastError.value = null
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val message = gson.fromJson(text, WebSocketMessage::class.java)
                    messagesChannel.trySend(message)
                } catch (e: Exception) {
                    _lastError.value = "Failed to parse message: ${e.message}"
                }
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                try {
                    val text = bytes.utf8()
                    val message = gson.fromJson(text, WebSocketMessage::class.java)
                    messagesChannel.trySend(message)
                } catch (e: Exception) {
                    _lastError.value = "Failed to parse message: ${e.message}"
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                _connectionState.value = WebSocketConnectionState.DISCONNECTED
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _connectionState.value = WebSocketConnectionState.DISCONNECTED
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                _connectionState.value = WebSocketConnectionState.DISCONNECTED
                _lastError.value = t.message
            }
        })
    }

    fun disconnect() {
        webSocket?.close(1000, "Normal closure")
        webSocket = null
        _connectionState.value = WebSocketConnectionState.DISCONNECTED
    }

    fun sendMessage(message: WebSocketMessage) {
        if (_connectionState.value != WebSocketConnectionState.CONNECTED) {
            _lastError.value = "WebSocket not connected"
            return
        }

        try {
            val json = gson.toJson(message)
            webSocket?.send(json)
        } catch (e: Exception) {
            _lastError.value = "Failed to send message: ${e.message}"
        }
    }

    // Convenience methods
    fun joinRoom(roomId: String) {
        val playerId = authManager.getCurrentUserId() ?: return
        val message = WebSocketMessage(
            type = "join_room",
            roomId = roomId,
            playerId = playerId,
            data = null,
            timestamp = System.currentTimeMillis()
        )
        sendMessage(message)
    }

    fun leaveRoom(roomId: String) {
        val playerId = authManager.getCurrentUserId() ?: return
        val message = WebSocketMessage(
            type = "leave_room",
            roomId = roomId,
            playerId = playerId,
            data = null,
            timestamp = System.currentTimeMillis()
        )
        sendMessage(message)
    }

    fun sendGameMove(roomId: String, moveData: Any) {
        val playerId = authManager.getCurrentUserId() ?: return
        val message = WebSocketMessage(
            type = "game_move",
            roomId = roomId,
            playerId = playerId,
            data = gson.toJson(moveData),
            timestamp = System.currentTimeMillis()
        )
        sendMessage(message)
    }

    fun sendChatMessage(roomId: String, text: String) {
        val playerId = authManager.getCurrentUserId() ?: return
        val chatData = mapOf("message" to text)
        val message = WebSocketMessage(
            type = "chat_message",
            roomId = roomId,
            playerId = playerId,
            data = gson.toJson(chatData),
            timestamp = System.currentTimeMillis()
        )
        sendMessage(message)
    }
}

enum class WebSocketConnectionState {
    DISCONNECTED,
    CONNECTING,
    CONNECTED
}